helix-term/src/commands.rs

```diff
@@ -3993,6 +3949,226 @@ pub fn completion(cx: &mut Context) {
         },
     );
 }
+fn completion_word(cx: &mut Context) {
+    completion_word_impl(cx, true)
+}
+fn completion_word_impl(cx: &mut Context, search_root: bool) {
+    let (view, doc) = current!(cx.editor);
+    let text = doc.text().slice(..);
+    let cursor = doc.selection(view.id).primary().cursor(text);
+    use helix_core::chars;
+    let mut iter = text.chars_at(cursor);
+    iter.reverse();
+    let offset = iter.take_while(|ch| chars::char_is_word(*ch)).count();
+    let start_offset = cursor.saturating_sub(offset);
+    let prefix = regex::escape(&text.slice(start_offset..cursor).to_string());
+
+    completion_grep(
+        cx,
+        format!(r"\b{}\w+", prefix,),
+        start_offset,
+        search_root,
+        false,
+    );
+}
+
+fn completion_line(cx: &mut Context) {
+    let (view, doc) = current!(cx.editor);
+    let text = doc.text().slice(..);
+    let cursor = doc.selection(view.id).primary().cursor(text);
+    let line_start = text.line_to_byte(text.byte_to_line(cursor));
+    use helix_core::chars;
+    let iter = text.chars_at(line_start);
+    let offset = iter.take_while(|ch| chars::char_is_whitespace(*ch)).count();
+    let start_offset = line_start + offset;
+    let prefix = regex::escape(&text.slice(start_offset..cursor).to_string());
+
+    completion_grep(cx, format!(r"^\s*{}.+", prefix), start_offset, true, true);
+}
+fn completion_grep(
+    cx: &mut Context,
+    query: String,
+    start_offset: usize,
+    search_root: bool,
+    match_whold_line: bool,
+) {
+    let (view, doc) = current!(cx.editor);
+    let text = doc.text().slice(..);
+    let current_content = text.to_string();
+    let cursor = doc.selection(view.id).primary().cursor(text);
+    let trigger_offset = cursor;
+    let config = cx.editor.config().file_picker.clone();
+
+    let callback = Box::pin(async move {
+        let matches = grep_impl(
+            &query,
+            current_content,
+            if search_root { Some(config) } else { None },
+            match_whold_line,
+        )
```

Comment:
I think we either want to only use open buffers or current buffer. Scanning the entire tree can easily be very costly and tons of disk I/O.

Reply:
It's very useful for editing source code without lsp, and completing lines is also very useful for writing similar logic in other files even there is a lsp.

We don't scan the file tree for C-x or idle completions, only do it when manually trigger by C-n and C-l, but also limit the match count to a few number.